src/main/java/com/mongodb/kafka/connect/util/jmx/internal/MongoMBean.java

```diff
@@ -0,0 +1,142 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.mongodb.kafka.connect.util.jmx.internal;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Consumer;
+import javax.management.Attribute;
+import javax.management.AttributeList;
+import javax.management.AttributeNotFoundException;
+import javax.management.DynamicMBean;
+import javax.management.MBeanAttributeInfo;
+import javax.management.MBeanInfo;
+
+public class MongoMBean implements DynamicMBean {
+  protected static final long[] MS_LIMITS = new long[] {1, 10, 100, 1000, 10000};
+
+  private String mBeanName;
+  private final List<Metric> metrics = new ArrayList<>();
+  private final Map<String, MetricValue> metricsMap = new HashMap<>();
+
+  protected MongoMBean(final String mBeanName) {
+    this.mBeanName = mBeanName;
+  }
+
+  public String getName() {
+    return mBeanName;
+  }
+
+  protected TotalMetric registerTotal(final String name) {
+    return register(new TotalMetric(name));
+  }
+
+  protected LatestMetric registerLatest(final String name) {
+    return register(new LatestMetric(name));
+  }
+
+  protected AdditiveMetric registerMs(final String name) {
+    return register(new AdditiveMetric(name, "ms", MS_LIMITS));
+  }
+
+  protected <T extends Metric> T register(final T m) {
+    metrics.add(m);
+    m.emit(value -> metricsMap.put(value.getName(), value));
+    return m;
+  }
+
+  public void emit(final Consumer<MetricValue> consumer) {
+    this.metrics.forEach(x -> x.emit(consumer));
+  }
+
+  public String toJSON() {
+    StringBuilder sb = new StringBuilder("{");
+    emit(
+        (v) -> {
+          if (sb.length() > 1) {
+            sb.append(", ");
+          }
+          sb.append("\"").append(v.getName()).append("\": ");
+          sb.append(v.get());
+        });
+    sb.append("}");
+    return sb.toString();
+  }
+
+  @Override
+  public Object getAttribute(final String name) throws AttributeNotFoundException {
+    if (metricsMap.containsKey(name)) {
+      return metricsMap.get(name).get();
+    } else {
+      throw new AttributeNotFoundException("getAttribute failed: value not found for: " + name);
+    }
+  }
+
+  @Override
+  public AttributeList getAttributes(final String[] attributes) {
+    AttributeList list = new AttributeList();
+    for (String name : attributes) {
+      if (metricsMap.containsKey(name)) {
+        list.add(metricsMap.get(name).get());
+      }
+    }
+    return list;
+  }
+
+  @Override
+  public MBeanInfo getMBeanInfo() {
+    List<MBeanAttributeInfo> attrs = new ArrayList<>();
+    for (Metric metric : this.metrics) {
```

Comment:
What's weird is that the DynamicMBean Javadoc doesn't state this requirement.

Reply:
In the current code, the MBean never changes. Each metric is registered with the MBean on field initialization, before the MBean is registered (and therefore visible for an `getMBeanInfo` invocation). This registration pattern is similar to [what Kafka does](https://github.com/a0x8o/kafka/blob/master/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/ConnectMetricsRegistry.java#L385), and allows fields to be specified as follows:
```
  private Metric failedRecords = registerTotal("records-failed");
  private Metric latestOffsetMs = registerLatest("latest-offset-ms");
  private Metric taskInvocations = registerMs("task-invocations");
```
It is true that we might later decide to register a new metric after the MBean has been constructed, and inadvertently violate this constraint, but this seems improbable, and difficult to prevent in code regardless.

Our tests already assert that each MBean produces a complete set of all metric attributes, after the test completes. I can add a test that confirms that all metric attributes available immediately after the MBean constructor is invoked match that complete set of metric attributes. I think this test would be more explicit about the constraint, and would cover everything we cannot cover in code. Would this address your concern?

Reply:
sgtm